onnxruntime/core/session/inference_session.h

```diff
@@ -328,6 +328,8 @@ class InferenceSession {
                             const OrtValueCachePtr& cache);
 #endif
 
+  void Replay();
```

Comment:
`Replay()` seems to be a CUDA-specific API and hence doesn't really fit in `InferenceSession` as a generic API. I would have thought `Replay()` is really part of `Run()` when the provider is the CUDA EP and when using cuda graph is enabled by the user.

Reply:
That's a good point! Here are some of my understandings. `Replay()` depends on the captured CUDA graph. That means users need to first generate the captured CUDA graph in order to run the inference, which makes the logic a little different from the current logic of `Run()`. 

1. Follow your suggestions (if my understanding is correct), maybe we can put the logic of both CUDA graph capturing  and replay into `Run()`. This way will not require any new public APIs, but need to add more logics to the current backend code. 
```
if (enable_cuda_graph && is_not_captured) {
  CaptureCudaGraphBegin();
  CurrentRun();
  CaptureCudaGraphEnd();
} else if (enable_cuda_graph && is_captured) {
  Replay();
} else {
  CurrentRun();
}
```

2. Another way is to create Python binding for CudaGraph class. This way will make it much easier for users to control CUDA graphs (e.g., create multiple CUDA graphs for different cases, edit CUDA graph, set different streams for `Replay()`). And the API logic seems to be more transparent.
```
cuda_graph = sess.capture_cuda_graph(...)
cuda_graph.replay()
```

Either way will work for me. What do you think of the above solutions? or any other suggestions?

Reply:
My thought is to avoid introducing public APIs wherever possible and staying away from introducing a new python binding class to begin with (we can always add it later but adding it now would mean we commit to long-term maintenance).

I have thought about it a bit more and here is the raw idea (a bit like your idea 1):

1) I would add a provider option to configure the CUDA EP to tell it to use cuda graphs. See here for an example of configuring an EP (`do_copy_in_default_stream`) in Python: https://github.com/microsoft/onnxruntime/blob/baa17679221807ea332a7134c0d10b81adf4009e/onnxruntime/test/python/onnxruntime_test_python.py#L1033. You may pick an option similar to the one being added in #10423 (like `enable_cuda_graph`). One could turn on the stream capture in CUDAExecutionProvider::OnRunStart() and stop the capture in CUDAExecutionProvider::OnRunEnd() iff the CUDA EP is configured to use cuda graph and if no graph has been captured thus far. Alternatively, an RunOption to give the user more control as to which Run will be captured as a graph could be a nice enhancement but it doesn't have to be in this PR.

2) In Run(), if the CUDA provider is part of the provider list AND if it is configured to use cuda graphs AND if a graph has already been captured, then we invoke provider->Replay() and skip all the downstream parts of Run() - OnRunStart(), ExecuteGraph(), OnRunEnd(), etc.

What do you think about this ?

Reply:
Yeah, I think this is a good solution: avoid introducing the public APIs and easy to use. For `OnRunStart()` and `OnRunEnd()`, just a small thing to add: there may be some across-device data transfer between them ([here](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/utils.cc#L666)), which may break CUDA graph capturing. It should be OK if we feed the inputs and outputs as CUDA values by using `run_with_iobinding(...)`.

I agree with this solution. Once it is finalized, I will update the PR accordingly.

Reply:
Not quite following. If/When cross-device copies are made during `ExecuteGraphImpl()` when the graph is getting captured, doesn't it use the same stream in the CUDA EP to make the copy ? If so, shouldn't that get captured during the graph capture too ? So, if the users use the same virtual memory address to pass in data during the graph replay, shouldn't the copy happen during the replay too ? It is possible I didn't understand some nuance of cuda graphs. Please correct me if I am wrong.

Reply:
If I remember correctly, the root cause is the synchronization like [here](https://github.com/microsoft/onnxruntime/blob/baa17679221807ea332a7134c0d10b81adf4009e/onnxruntime/core/providers/cuda/gpu_data_transfer.cc#L61). As mentioned in the [CUDA doc](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs), `It is therefore also invalid to call synchronous APIs in this case.`

Reply:
I have tried out an interface in https://github.com/microsoft/onnxruntime/pull/10459. Curious to get your thoughts on that approach.

Reply:
The proposed APIs have been integrated into this PR. Marked is as resolved but please feel free to reopen it.